### PR TITLE
add path sum of binary tree

### PR DESCRIPTION
**Pull request by iattempt**
**Fixes issue:**

**Changes:**
1. add path sum (whole and part path) of binary tree
2. add tests

### CODE REVIEW

File: code/data_structures/tree/path_sum/path_sum.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+// private
+
+template<typename _Ty, typename _Compare, class _TreeNode>
+void
+PathSum<_Ty, _Compare, _TreeNode>::getPathsOfSumUp(PNode<_Ty> root,
+                                                   std::vector<_Ty> prev,
+                                                   _Ty prev_sum,
+                                                   _Ty const &sum,
+                                                   std::vector<std::vector<_Ty>> &res) {
+    if (root!=nullptr)
```
Comment (arnavb) [inline code/data_structures/tree/path_sum/path_sum.cpp]: Add spacing between the operands/operator.

File: code/data_structures/tree/path_sum/path_sum.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+{
+private:
+    template<typename _T>
+    using Node = _TreeNode;
+    template<typename _T>
+    using PNode = std::shared_ptr<Node<_T>>;
+    using size_type = size_t;
+
+public:
+    enum class PathType {Whole, Part};
```
Comment (arnavb) [inline code/data_structures/tree/path_sum/path_sum.hpp]: Whenever using braces, always have a space after the opening brace, and before the closing one.

File: code/data_structures/tree/path_sum/path_sum.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    PathType path_type_;
+
+    void getPathsOfSumUp(PNode<_Ty> root,
+                         std::vector<_Ty> prev,
+                         _Ty prev_sum,
+                         _Ty const &sum,
+                         std::vector<std::vector<_Ty>> &res);
+};
+
+#include "path_sum.cpp"
```
Comment (arnavb) [inline code/data_structures/tree/path_sum/path_sum.hpp]: Why?

Comment (iattempt) [reply code/data_structures/tree/path_sum/path_sum.hpp]: refer to https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file

**Comment by iattempt** [comment]
thanks for review~👍

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**